wire-ios-data-model/Source/MLS/OneOnOne/OneOnOneResolver.swift

```diff
@@ -92,68 +91,151 @@ public final class OneOnOneResolver: OneOnOneResolverInterface {
+        in context: NSManagedObjectContext
+    ) async -> OneOnOneConversationResolution {
+        WireLogger.conversation.debug("no common protocols found")
+
+        await context.perform {
+            guard
+                let otherUser = ZMUser.fetch(with: userID, in: context),
+                let conversation = otherUser.oneOnOneConversation
+            else {
+                return
+            }
 
-                conversation.isForcedReadOnly = true
+            self.makeConversationReadOnly(
+                selfUser: ZMUser.selfUser(in: context),
+                otherUser: otherUser,
+                conversation: conversation
+            )
+        }
 
-            }
-            return .archivedAsReadOnly
+        return .archivedAsReadOnly
+    }
 
-        case .mls:
-            WireLogger.conversation.debug("should resolve to mls 1-1 conversation")
+    private func makeConversationReadOnly(
+        selfUser: ZMUser,
+        otherUser: ZMUser,
+        conversation: ZMConversation
+    ) {
+        if conversation.isForcedReadOnly { return }
 
-            guard let migrator else {
-                throw OneOnOneResolverError.migratorNotFound
-            }
+        if !selfUser.supportedProtocols.contains(.mls) {
+            conversation.appendMLSMigrationMLSNotSupportedForSelfUser(user: selfUser)
+        } else if !otherUser.supportedProtocols.contains(.mls) {
+            conversation.appendMLSMigrationMLSNotSupportedForOtherUser(user: otherUser)
+        }
+
+        conversation.isForcedReadOnly = true
+    }
+
+    // MARK: Resolve - MLS
+
+    private func resolveCommonUserProtocolMLS(
+        with userID: QualifiedID,
+        in context: NSManagedObjectContext
+    ) async throws -> OneOnOneConversationResolution {
+        WireLogger.conversation.debug("should resolve to mls 1-1 conversation")
 
-            let mlsGroupIdentifier = try await migrator.migrateToMLS(
+        guard let migrator else {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        guard let mlsService else {
+            throw OneOnOneResolverError.migratorNotFound
+        }
+
+        let mlsGroupID = try await syncMLSConversationFromBackend(
+            userID: userID,
+            in: context
+        )
+
+        if await mlsService.conversationExists(groupID: mlsGroupID) {
+            return .noAction
+        }
+
+        guard let epoch = await fetchMLSConversationEpoch(mlsGroupID: mlsGroupID, in: context) else {
+            throw MigrateMLSOneOnOneConversationError.missingConversationEpoch
+        }
+
+        if epoch == 0 {
+            // migrate to a new conversation
+            try await migrator.migrateToMLS(
                 userID: userID,
+                mlsGroupID: mlsGroupID,
                 in: context
             )
+        } else {
+            // join existing conversation via external commit
+            try await mlsService.joinGroup(with: mlsGroupID)
```

Comment:
When looking at `migrateToMLS` above, migrating to MLS if epoch is 0 involves two steps:
- Establishing the MLS group  (in `establishLocalMLSConversationIfNeeded`)
- Switching from the proteus `ZMConversation` to the mls `ZMConversation` (in `switchLocalConversationToMLS`)

In the case where epoch is > 0, wouldn't we only join the MLS group but miss switching from the proteus conversation to the mls conversation?

Reply:
Can you please describe a case where this happen? I thought either:
- we have a proteus conversation the epoch is always 0 when mls conversation is created, then the migration is executed
- we have an existing mls conversation, so the epoch is > 0 and we just join with the new device

Reply:
The epoch being zero or not is not dependent upon there being an existing  proteus conversation.

If epoch > 0 some else already established the MLS group and it already exists on the backend:
    1. Join MLS group via external commit
    2. Set `oneOnOneConversation` to MLS 1-1 conversations
    3. Migrate messages from proteus 1-1 conversation

if epoch == 0 MLS group doesn't exist on backend so we need to create it
  1. Establish MLS group with other user
  2. Set `oneOnOneConversation` to MLS 1-1 conversations
  3. Migrate messages from proteus 1-1 conversation

The only difference between the two cases is what we do in step 1.

Reply:
Thanks, that helped. With David I discussed it similar yesterday and did the changes here: a5b2e7beb6da359e4e1597f14a69365aeeaec213

And there are more things to consider:
- if the epoch can not be found in core data locally, there is another mistake that e.g. the preconditions to sync the conversation from the backend didn't work properly
- when no proteus conversation exists, because the user login on the new device the first time, `step 3` must be skipped (instead of throwing an error, because the conversation can not be found etc).